src/libsync/vfs/cfapi/cfapiwrapper.cpp

```diff
@@ -124,21 +124,75 @@ void CALLBACK cfApiFetchDataCallback(const CF_CALLBACK_INFO *callbackInfo, const
-        auto data = socket.readAll();
+        const auto data = socket.readAll();
         if (data.isEmpty()) {
             qCWarning(lcCfApiWrapper) << "Unexpected empty data received" << requestId;
             sendTransferError();
             loop.quit();
             return;
         }
-        sendTransferInfo(data, offset);
-        offset += data.length();
+
+        QByteArray dataToSent;
+        // CFAPI expectes sended blocks to be of a specified size.
+        // Only the last sended block is allowed to be smaller than the specified block size
+        // and therefore marks the end of the transmission.
+        static constexpr auto CFAPI_BLOCK_SIZE = 4096;
+
+        int dataIndex = 0;
+        // Is there any data left from last run?
+        if (!protrudingData.isEmpty()) {
+            QByteArray dataToSent;
+            dataToSent.append(protrudingData);
+
+            // Fill up remaining first block if possible
+            const auto fillupSize = CFAPI_BLOCK_SIZE - dataToSent.size();
+            if (fillupSize > 0) {
+                dataToSent.append(data.left(fillupSize));
+                dataIndex += fillupSize;
+            }
+
+            // Send the block
+            qDebug(lcCfApiWrapper) << "Send protruding data that is left from last run. Size:" << dataToSent.size();
+            sendTransferInfo(dataToSent, offset);
+            offset += dataToSent.length();
+        }
+
+        // Do we still have data to sent?
+        if (dataIndex < data.size()) {
+            const auto blocksCount = (data.size() - (dataIndex + 1)) / CFAPI_BLOCK_SIZE;
+            if (blocksCount != 0) {
+                // Send the blocks one by one
+                for (int i = 0; i < blocksCount; ++i) {
+                    auto dataToSent = data.mid(dataIndex, CFAPI_BLOCK_SIZE);
+                    qDebug(lcCfApiWrapper) << "Send data block. Size:" << dataToSent.size();
+                    sendTransferInfo(dataToSent, offset);
+                    offset += dataToSent.length();
+                    dataIndex += CFAPI_BLOCK_SIZE;
+                }
+            }
+
+            // Do we still have data left? If yes, save this data for the next run and send it then
+            protrudingDataLeft = false;
+            if (dataIndex < data.size()) {
+                protrudingDataLeft = true;
+                protrudingData = data.mid(dataIndex, data.size() - dataIndex);
+                qDebug(lcCfApiWrapper) << "Save protruding data for next run. Size:" << protrudingData.size();
+            }
+        }
     });
 
     QObject::connect(vfs, &OCC::VfsCfApi::hydrationRequestFinished, &loop, [&](const QString &id, int s) {
+        qDebug(lcCfApiWrapper) << "Hydration finished for request" << id;
         if (requestId == id) {
+            if (protrudingDataLeft) {
+                qDebug(lcCfApiWrapper) << "Send protruding data left. Size:" << protrudingData.size();
+                sendTransferInfo(protrudingData, offset);
```

Comment:
@FlexW Overall, I'd say, if sending multiples of 4096 works similarly to sending single 4096-sized blocks, you can even end up having only one **QByteArray** buffered that is getting checked if it has enough data for even a single 4096 block, and then, rest of data is just kept inside it after 4096-multiple blocks are sent. Then, the upcoming data will just get appended to that single bucket. While the bucket does not have enough data for at least one 4096-sized block, it would keep getting filled-in. Everytime it reaches the state when it has at least one 4096 blocks - it gets flashed to a `sendTransferInfo` and, whatever is left that is protruding, can then be assumed as the bottom of the bucket (or a head of a QByteArray).

The size of **data** wouldn't exceed something like 4096 * 4 most of the time, so, we shouldn't expect **QByteArray** to grow too big. The capacity will be increasing each time it gets bigger than the size, and later on, what's been allocated will just be re-used. So, we shouldn't have performance drop.

@er-vin Correct me if I am wrong. This is how I understood your idea with a bucket that we are feeding with data until we have aligned blocks that we can send.